### PR TITLE
Improve code coverage for Bucket operations.

### PR DESCRIPTION
**Pull request by coryan**
We were no doing any integration tests for failures, we should at least
verify that the exceptions raised match our minimal expectations.

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/1206)

### CODE REVIEW

**Comment by coryan** [comment]
Rebased to resolve conflicts. PTAL.

File: google/cloud/storage/testbench/testbench.py [hunk 1]
```diff
@@ -1345,6 +1350,9 @@ def buckets_insert():
     if bucket_name is None:
         raise error_response.ErrorResponse(
             'Missing bucket name in `Buckets: insert`', status_code=412)
+    if len(bucket_name) > 63 or not re.match('^[a-z][a-z-]*[a-z]', bucket_name):
```
Comment (houglum) [inline google/cloud/storage/testbench/testbench.py]: I think this regex is incorrect, given the guidelines at https://cloud.google.com/storage/docs/naming.  I actually had a snippet laying around that covers the checks mentioned in our docs, if you'd rather use that (probably worth factoring out into its own method):

```
valid = True
if '.' in bucket_name:
    valid &= (
        len(bucket_name) <= 222 and
        all([len(part) <= 63 for part in bucket_name.split('.')]))
else:
    valid &= len(bucket_name) <= 63
valid &= re.match('^[a-z0-9][a-z0-9._-]+[a-z0-9]$', bucket_name)
valid &= not bucket_name.startswith('goog')
valid &= re.search('g[0o][0o]g[1l][e3]', bucket_name)
valid &= re.match('^[0-9]{1,3}[.][0-9]{1,3}[.][0-9]{1,3}[.][0-9]{1,3}$', bucket_name)
```

The snippet above allows underscores -- while we discourage their use, they're still technically allowed in (GCS) bucket names.

File: google/cloud/storage/tests/bucket_integration_test.cc [hunk 2]
```diff
@@ -559,6 +559,194 @@ TEST_F(BucketIntegrationTest, IamCRUD) {
...
+#else
+  EXPECT_DEATH_IF_SUPPORTED(results.assign(stream.begin(), stream.end()),
+                            "exceptions are disabled");
+#endif  // GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
+}
+
+TEST_F(BucketIntegrationTest, CreateFailure) {
+  Client client;
+
+  // Try to create an invalid bucket (the name should not contain `_`), the
```
Comment (houglum) [inline google/cloud/storage/tests/bucket_integration_test.cc]: This is true for s3 buckets, but not GCS buckets, as mentioned above

File: google/cloud/storage/testbench/testbench.py (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/testbench/testbench.py]: Done.

File: google/cloud/storage/tests/bucket_integration_test.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/tests/bucket_integration_test.cc]: Fixed.

**Comment by coryan** [comment]
PTAL.

File: google/cloud/storage/testbench/testbench_utils.py [hunk 3]
```diff
@@ -18,6 +18,35 @@
...
+    :param bucket_name:str the name to validate.
+    :rtype: bool
+    """
+    valid = True
+    if '.' in bucket_name:
+        valid &= len(bucket_name) <= 222
+        valid &= all([len(part) <= 63 for part in bucket_name.split('.')])
+    else:
+        valid &= len(bucket_name) <= 63
+    valid &= re.match('^[a-z0-9][a-z0-9._\\-]*[a-z0-9]$', bucket_name) is not None
```
Comment (houglum) [inline google/cloud/storage/testbench/testbench_utils.py]: Nit: This `*` should be a `+`, since bucket names must be at least 3 chars long.

**Pull request merged**